### PR TITLE
New ReCapSegCaller and CnvCaller classes:  ReCapSeg caller is near-identical match to python caller.  Addressing some inconsistencies that were missed in the unit tests (s…

### LINKED ISSUE #136: Log more information from the caller
Currently, it is very hard to determine the mode and threshold boundaries that the caller chose for a particular sample.  

Solution:

Add log messages detailing these.

### LINKED ISSUE #137: Make caller use the segment mean instead of the mean of the target CR estimates.
Since the caller insists on ModeledSegment, there should always be a segment mean.

### LINKED ISSUE #144: Caller has (seemingly) random dropoff in sensitivity
This is due to the standard deviation being influenced by outlier targets in some samples.  See the purity 70% (SM-74P3K-70) in the graph below.

ampsensitivity

### LINKED ISSUE #151: Create a caller that mimics the python caller exactly and make it the default.

### PR DESCRIPTION
**Pull request by LeeTL1220**
…o that the tests passed).  Added logging messages and doc improvements.  Made the caller use the segment mean.

Updated version of GATK that is used by HB-p

Updated version of GATK to incorporate last minute changes to exome coverage collection.

Fixing gradle wrapper version number

Filtering the target for outliers before calculating the sigma.  Minor formatting change.

Making Java code closer to match with python code.

Pruning outliers using a percentile to detect prune the low value targets (which is the more common error mode)

Added legacy flag for reading in older seg files into the caller.

Made the default caller a replica of the python caller and then added a new caller class for new caller functionality.

Added unit tests to test the CnvCaller

Extended unit tests to hit the CnvCaller as well.

Closes #151 
Closes #144 
Closes #137 
Closes #136

### CODE REVIEW

**Comment by LeeTL1220** [comment]
@davidbenjamin Sigh... found issue.  I will reassign when ready.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/CnvCaller.java [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+        // Get the standard deviation of targets belonging to high-confidence neutral segments
+        final double[] nearNeutralCoverage = segments.stream()
+                .filter(s -> Math.abs(s.getSegmentMean() - neutralCoverage) < COPY_NEUTRAL_CUTOFF)
+                .flatMap(s -> targets.targets(s).stream())
+                .mapToDouble(TargetCoverage::getCoverage)
+                .toArray();
+        final double neutralSigma = new StandardDeviation().evaluate(nearNeutralCoverage);
+        logger.info(String.format("Neutral Sigma (unfiltered): %.4f", neutralSigma));
+
+        final double filterThreshold = new Percentile(2).evaluate(nearNeutralCoverage);
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/CnvCaller.java]: Magic constant

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ModeledSegment.java [hunk 2]
```diff
@@ -24,10 +24,10 @@
 
     private long originalProbeCount;
 
-    public ModeledSegment(final SimpleInterval interval, final String call, final long originalProbeCount, final double segmentMean) {
+    public ModeledSegment(final SimpleInterval interval, final String call, final long originalProbeCount, final double segmentMeanInLogCR) {
         this.simpleInterval = Utils.nonNull(interval, "The input interval cannot be null");
         this.call = Utils.nonNull(call, String.format("The input call cannot be null.  Use empty string, instead (\"%s\")", NO_CALL));
-        this.segmentMean = ParamUtils.isFinite(segmentMean, "Segment Mean must be a finite, real number.");
+        this.segmentMean = ParamUtils.isFinite(segmentMeanInLogCR, "Segment Mean must be a finite.");
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ModeledSegment.java]: Extraneous 'a'

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ReCapSegCaller.java [hunk 3]
```diff
@@ -1,92 +1,97 @@
...
 
     //bounds on log_2 coverage for high-confidence neutral segments
     public static final double COPY_NEUTRAL_CUTOFF = 0.1;
 
     private ReCapSegCaller() {} // prevent instantiation
 
+    private static double calculateT(final TargetCollection<TargetCoverage> targets, final List<ModeledSegment> segments) {
+
+        //Get the segments that are likely copy neutral.
+        // Math.abs removed to mimic python...
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ReCapSegCaller.java]: Doesn't the lack of abs mean that deletions are considered neutral?

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ReCapSegCaller.java [hunk 4]
```diff
@@ -1,92 +1,97 @@
...
+        final List<TargetCoverage> copyNeutralTargets = copyNeutralSegments.stream()
+                .flatMap(s -> targets.targets(s).stream()).collect(Collectors.toList());
+
+        final double [] copyNeutralTargetsCopyRatio = copyNeutralTargets.stream().mapToDouble(TargetCoverage::getCoverage).toArray();
+        final double meanCopyNeutralTargets = new Mean().evaluate(copyNeutralTargetsCopyRatio);
+        final double sigmaCopyNeutralTargets = new StandardDeviation().evaluate(copyNeutralTargetsCopyRatio);
+
+        // Now we filter outliers by only including those w/in 2 standard deviations.
+        final double [] filteredCopyNeutralTargetsCopyRatio = Arrays.stream(copyNeutralTargetsCopyRatio)
+                .filter(c -> Math.abs(c - meanCopyNeutralTargets) < sigmaCopyNeutralTargets * 2).toArray();
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ReCapSegCaller.java]: I would prefer a named constant for the Z-score threshold.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ReCapSegCaller.java [hunk 5]
```diff
@@ -1,92 +1,97 @@
...
-            //Get the number of standard deviations the mean falls from the copy neutral value
-            //we should really use the standard deviation of the mean of segment.numTargets() targets
-            //which is sigma/ sqrt(numTargets).  However, the underlying segment means vary due to noise not
-            //removed by tangent normalization, and such a caller would be too strict.
-            //Thus we just use the population standard deviation of targets, and defer a mathematically
-            //principled treatment for a later caller
-            final double Z = (SegmentUtils.meanTargetCoverage(segment, targets) - neutralCoverage)/neutralSigma;
 
-            final String call = Z < -zThreshold ? DELETION_CALL: (Z > zThreshold ? AMPLIFICATION_CALL : NEUTRAL_CALL);
+            String call = NEUTRAL_CALL;
```
Comment (davidbenjamin) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ReCapSegCaller.java]: Haha we have different affection for ternaries.  I don't object to your preference.

**Comment by davidbenjamin** [comment]
@LeeTL1220 Review complete.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ReCapSegCaller.java (hunk 3 above)
Comment (LeeTL1220) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ReCapSegCaller.java]: Yup ... the comment says it all...  I think this was a bug in the python.
No action.

File: src/main/java/org/broadinstitute/hellbender/tools/exome/CnvCaller.java (hunk 1 above)
Comment (LeeTL1220) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/CnvCaller.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ModeledSegment.java (hunk 2 above)
Comment (LeeTL1220) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ModeledSegment.java]: Done

File: src/main/java/org/broadinstitute/hellbender/tools/exome/ReCapSegCaller.java (hunk 4 above)
Comment (LeeTL1220) [inline src/main/java/org/broadinstitute/hellbender/tools/exome/ReCapSegCaller.java]: Done

**Comment by LeeTL1220** [comment]
@davidbenjamin Addressed comments

**Comment by davidbenjamin** [comment]
@LeeTL1220 I will merge this after Travis finishes.

**Pull request merged**